content/blog/quine-quine/index.md

```diff
@@ -0,0 +1,227 @@
+- It must not be an empty program
+
+## A first attempt
+
+If we had this simple Ruby program:
+
+```ruby
+puts "hello"
+```
+
+And we executed it, we would get:
+
+```bash
+$ ruby quine.rb
+hello
+```
+
+So we could make our program to look like this:
+
+```ruby
+puts "puts \"hello\""
+```
+
+But running that would only yield:
+
+```bash
+$ ruby quine.rb
+puts "hello"
+```
+
+As we can see, this approach won't get us closer to the solution since
+we're always going to miss the outer `puts`.
+
+## A bit of self reflection
+
+Luckily in Ruby there is a simple way to escape this using
+`Object#inspect`:
+
+```ruby
+src = "puts src"
+puts "src = " + src.inspect
+```
+
+```bash
+$ ruby quine.rb
+src = "puts src"
+```
+
+But we're still missing the last line. What if we added the `src` again
+to the end of our `puts` statement, but this time as a normal string?
+
+```ruby
+src = "puts src"
+puts "src = " + src.inspect + src
+```
+
+```bash
+$ ruby quine.rb
+src = "puts src"puts src
+```
+
+Uh! Since the contents of the `src` string don't really matter to the
+behaviour of our program, we can get quite creative with it!
+
+```ruby
+src = "\nputs \"src = \" + src.inspect + src"
+puts "src = " + src.inspect + src
+```
+
+```bash
+$ ruby quine.rb
+src = "\nputs \"src = \" + src.inspect + src"
+puts "src = " + src.inspect + src
+```
+
+Tada! 🎉
+
+## How to write your own Quine
+
+This simple example showed us some of the core ideas for writing a Quine:
+
+- We build the program in two parts, one that we call the _code_ and one
+  that we call the _data_. The data part of the program is derived from the
+  code part, usually by some means of escaping.
+- When executed, the code first uses the data to print the data.
+- Then the code uses the data to print the code. This is generally easy
+  because we intentionally derived the data from the code.
+
+A great analogy of a Quine comes from cellular biology: if we think of the
+cell as the **code** and the cell's DNA as the **data**, then we'll see that
+the cell is able to replicate itself using the DNA, and this process
+involves replicating the DNA itself. On the other hand the DNA contains
+all the information for replicating the cell, but without the cell itself
+it would just be an inert piece of data.
+
+## Quine Quine
+
+Douglas Hofstadter also invented the verb _to quine_, which means "to write
+something a first time, and then to write it a second time with quotation
+marks around it". For example, if we quine "say", we would get "say 'say'".
+If we were to quine "quine", we would get "quine 'quine'", which itself is
+a quine...
+
+## Testing, testing, testing
+
+When you're trying to write your own Quine, you'll get to a point when you
+think you've reached a solution, but end up having to compare two printouts
+which are not exactly human friendly. So you can use this bit of bash
+instead:
+
+```bash
+$ diff <(cat quine.rb) <(ruby quine.rb) && echo "IS QUINE"
```

Comment:
Since `quine.rb` is a file already, this could be simplified to:

```bash
$ diff quine.rb <(ruby quine.rb) && echo "IS QUINE"
```

Here and elsewhere.